### PR TITLE
[FEATURE] union alphabet and gapped alphabet with tests

### PR DESCRIPTION
**Pull request by eldariont**

### CODE REVIEW

File: include/seqan3/alphabet/union_alphabet.hpp [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+                }
+            }, *alphabets_it);
+
+            size_start = size_end;
+            sizes_it++;
+            alphabets_it++;
+        }
+
+        return result.value_or(static_cast<char_type>(0));
+    }
```
Comment (h-2) [inline include/seqan3/alphabet/union_alphabet.hpp]: This function is linear in the amount of alphabets and is quite complicated!

I would propose instead: create a bit vector that has a 1 at the beginning of every interval on the integral_value, then do a popcnt on the bitvector to find which of the i types you are in and call the appropriate to_char.

Comment (h-2) [inline include/seqan3/alphabet/union_alphabet.hpp]: could possibly be an sdsl::bit_vector, too.

File: include/seqan3/alphabet/union_alphabet.hpp [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+            sizes_it++;
+            alphabets_it++;
+        }
+
+        if(!found) {
+            value = 0;
+        }
+
+        return *this;
+    }
```
Comment (h-2) [inline include/seqan3/alphabet/union_alphabet.hpp]: respect for the metaprogramming, but couldn't this be done a little simpler? e.g. by recursing of the list of types and just check if `alphabet{}.from_char(c).to_char() == c`?

just a thought...

File: test/alphabet/union_alphabet.cpp [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+using namespace seqan3;
+
+TEST(union_alphabet_test, default_ctr)
+{
+    using alphabet_t = union_alphabet<dna4, gap>;
+    alphabet_t{};
+}
+
+// TODO: othe constructurs
```
Comment (h-2) [inline test/alphabet/union_alphabet.cpp]: ?

File: test/alphabet/union_alphabet.cpp [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    EXPECT_TRUE(expect1);
+
+    constexpr auto expect2 = std::is_same_v<alphabet2_t::integral_type, uint8_t>;
+    EXPECT_TRUE(expect2);
+
+    constexpr auto expect3 = std::is_same_v<alphabet3_t::integral_type, bool>;
+    EXPECT_TRUE(expect3);
+}
+
+// todo: constructors
```
Comment (h-2) [inline test/alphabet/union_alphabet.cpp]: ?

File: include/seqan3/alphabet/union_alphabet.hpp [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+template <typename tuple_t, size_t ...I>
+constexpr auto alphabet_objects(std::index_sequence<I...>)
+{
+    using variant_t = std::variant<std::tuple_element_t<I, tuple_t>...>;
+    using list_t = std::initializer_list<variant_t>;
+    return list_t{
+        variant_t{
+            std::in_place_index_t<I>{},
+            std::tuple_element_t<I, tuple_t>{}
+        }...
```
Comment (marehr) [inline include/seqan3/alphabet/union_alphabet.hpp]: nop

**Comment by h-2** [comment]
the algorithmic comments don't need to be addressed, if we keep a big TODO, but the others need to be addressed and we also need doxygen.

File: include/seqan3/alphabet/gap/gap.hpp [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+ */
+
+struct gap
+{
+    /* types */
+    //! the type of the alphabet when converted to char (e.g. via @link to_char @endlink)
+    using char_type = char;
+    //! the type of the alphabet when represented as a number (e.g. via @link to_integral @endlink)
+    using integral_type = bool;
+    using c_type = bool;
```
Comment (h-2) [inline include/seqan3/alphabet/gap/gap.hpp]: you can just use integral_type and get rid of c_type!

File: include/seqan3/alphabet/gap/gap.hpp [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+    //! return the letter as a character of @link char_type @endlink.
+    constexpr char_type to_char() const
+    {
+        return '-';
+    }
+
+    //! return the gap's numeric value.
+    constexpr integral_type to_integral() const
+    {
+        return static_cast<integral_type>(value);
```
Comment (h-2) [inline include/seqan3/alphabet/gap/gap.hpp]: just return 0 ?

Comment (h-2) [inline include/seqan3/alphabet/gap/gap.hpp]: or value

File: include/seqan3/alphabet/gap/gap.hpp [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+    //! assign from a character does not do anything.
+    constexpr gap & from_char(char_type const in)
+    {
+        return *this;
+    }
+
+    //! assign from a numeric value does not do anything.
+    constexpr gap & from_integral(integral_type const in)
+    {
+        return *this;
```
Comment (h-2) [inline include/seqan3/alphabet/gap/gap.hpp]: add an assert that the value is 0

File: include/seqan3/alphabet/gap/gap.hpp [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+        return value <= rhs.value;
+    }
+
+    constexpr bool operator>=(gap const & rhs) const
+    {
+        return value >= rhs.value;
+    }
+    //!@}
+};
+
```
Comment (h-2) [inline include/seqan3/alphabet/gap/gap.hpp]: wrap the static_assert in an `#ifndef NDEBUG` and also test the inner_alphabet type. See dna4 for more details

File: include/seqan3/alphabet/gap/gapped_alphabet.hpp [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    constexpr gapped_alphabet & from_char(typename union_alphabet<underlying_t, gap>::char_type const c)
+    {
+        union_alphabet<underlying_t, gap>::from_char(c);
+        return *this;
+    }
+};
+
+static_assert(alphabet_concept<gapped_alphabet<dna4>>);
```
Comment (h-2) [inline include/seqan3/alphabet/gap/gapped_alphabet.hpp]: wrap static_assert, see above

File: include/seqan3/alphabet/union_alphabet.hpp (hunk 1 above)
Comment (h-2) [inline include/seqan3/alphabet/union_alphabet.hpp]: even if you don't change the algorithmic stuff now, please correct the code style

File: include/seqan3/alphabet/union_alphabet.hpp [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+    }
+
+    constexpr bool operator>=(union_alphabet const & rhs) const
+    {
+        return value >= rhs.value;
+    }
+    //!@}
+};
+
+static_assert(detail::internal_alphabet_concept<union_alphabet<dna5, dna5>>);
```
Comment (h-2) [inline include/seqan3/alphabet/union_alphabet.hpp]: wrap in macro, see above

File: test/alphabet/gap/gap_test.cpp [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ==========================================================================
+// Author: David Heller <david.heller@fu-berlin.de>
+// ==========================================================================
+
+#include <seqan3/alphabet/gap/gap.hpp>
+#include <gtest/gtest.h>
+#include <sstream>
```
Comment (h-2) [inline test/alphabet/gap/gap_test.cpp]: includes in blocks, please! optimally sort alphabetically in the block :)

File: test/alphabet/gap/gapped_alphabet_test.cpp [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+// DAMAGE.
+//
+// ==========================================================================
+// Author: David Heller <david.heller@fu-berlin.de>
+// ==========================================================================
+
+#include <seqan3/alphabet/gap/gapped_alphabet.hpp>
+#include <seqan3/alphabet/nucleotide/dna4.hpp>
+#include <gtest/gtest.h>
+#include <sstream>
```
Comment (h-2) [inline test/alphabet/gap/gapped_alphabet_test.cpp]: see above

File: test/alphabet/union_alphabet.cpp [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+    using alphabet_t = union_alphabet<dna4, gap>;
+    alphabet_t letter1{};
+    alphabet_t letter2{std::move(letter1)};
+
+    EXPECT_EQ(letter1, letter2);
+}
+
+TEST(union_alphabet_test, assignment_union_base_types)
+{
+    // TODO:
```
Comment (h-2) [inline test/alphabet/union_alphabet.cpp]: ?

**Pull request merged**